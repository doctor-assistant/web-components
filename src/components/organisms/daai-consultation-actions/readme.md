# daai-consultation-actions



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute      | Description | Type      | Default     |
| -------------- | -------------- | ----------- | --------- | ----------- |
| `apikey`       | `apikey`       |             | `any`     | `""`        |
| `error`        | `error`        |             | `any`     | `undefined` |
| `event`        | `event`        |             | `any`     | `undefined` |
| `metadata`     | `metadata`     |             | `string`  | `undefined` |
| `professional` | `professional` |             | `string`  | `""`        |
| `specialty`    | `specialty`    |             | `any`     | `undefined` |
| `success`      | `success`      |             | `any`     | `undefined` |
| `telemedicine` | `telemedicine` |             | `boolean` | `undefined` |


## Dependencies

### Used by

 - [daai-consultation-recorder](../../templates/daai-consultation-recorder)

### Depends on

- [daai-button-with-icon](../../molecules/daai-button-with-icon)
- [daai-mic-icon](../../atoms/icons)
- [daai-text](../../atoms/text)
- [daai-menu-icon](../../atoms/icons)
- [daai-config](../../molecules/daai-config)
- [daai-resume-recording-icon](../../atoms/icons)
- [daai-pause-icon](../../atoms/icons)

### Graph
```mermaid
graph TD;
  daai-consultation-actions --> daai-button-with-icon
  daai-consultation-actions --> daai-mic-icon
  daai-consultation-actions --> daai-text
  daai-consultation-actions --> daai-menu-icon
  daai-consultation-actions --> daai-config
  daai-consultation-actions --> daai-resume-recording-icon
  daai-consultation-actions --> daai-pause-icon
  daai-button-with-icon --> daai-button
  daai-config --> daai-button-with-icon
  daai-config --> daai-config-mic-icon
  daai-config --> daai-text
  daai-config --> daai-menu-tutorial-icon
  daai-consultation-recorder --> daai-consultation-actions
  style daai-consultation-actions fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
