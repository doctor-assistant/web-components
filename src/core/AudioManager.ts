class AudioManager {
  private static instance: AudioManager;
  private context: AudioContext;
  private connections: Map<HTMLVideoElement, {
    source: MediaElementAudioSourceNode;
    gain: GainNode;
  }>;
  private microphoneStream: MediaStream | null = null;
  private videoStream: MediaStream | null = null;

  private constructor() {
    this.context = new AudioContext();
    this.connections = new Map();
  }

  static getInstance(): AudioManager {
    if (!AudioManager.instance) {
      AudioManager.instance = new AudioManager();
    }
    return AudioManager.instance;
  }

  private async connectWithRetry(videoElement: HTMLVideoElement, retries = 3): Promise<MediaStream> {
    for (let attempt = 0; attempt < retries; attempt++) {
      try {
        // Ensure we're disconnected before attempting connection
        this.disconnectVideoElement(videoElement);
        
        // Wait a bit between retries
        if (attempt > 0) {
          await new Promise(resolve => setTimeout(resolve, 100));
        }

        const source = this.context.createMediaElementSource(videoElement);
        const gain = this.context.createGain();
        const destination = this.context.createMediaStreamDestination();

        source.connect(gain);
        gain.connect(destination);

        this.connections.set(videoElement, { source, gain });
        this.videoStream = destination.stream;
        return destination.stream;
      } catch (error) {
        console.warn(`Connection attempt ${attempt + 1} failed:`, error);
        
        // On last retry, throw the error
        if (attempt === retries - 1) {
          throw error;
        }
      }
    }
    throw new Error('Failed to connect video element after all retries');
  }

  async connectVideoElement(videoElement: HTMLVideoElement): Promise<MediaStream> {
    try {
      return await this.connectWithRetry(videoElement);
    } catch (error) {
      console.error('Failed to connect video element:', error);
      // Ensure cleanup on failure
      this.disconnectVideoElement(videoElement);
      throw error;
    }
  }

  disconnectVideoElement(videoElement: HTMLVideoElement): void {
    const connection = this.connections.get(videoElement);
    if (connection) {
      try {
        connection.source.disconnect();
        connection.gain.disconnect();
      } catch (error) {
        console.warn('Error disconnecting video element:', error);
      }
      this.connections.delete(videoElement);
    }
  }

  async getMicrophoneStream(constraints: MediaTrackConstraints): Promise<MediaStream> {
    if (this.microphoneStream) {
      this.microphoneStream.getTracks().forEach(track => track.stop());
    }
    this.microphoneStream = await navigator.mediaDevices.getUserMedia({ audio: constraints });
    return this.microphoneStream;
  }

  createMixedStream(streams: MediaStream[]): MediaStream {
    const mixedStream = new MediaStream();
    streams.forEach(stream => {
      stream.getAudioTracks().forEach(track => mixedStream.addTrack(track));
    });
    return mixedStream;
  }

  cleanup(): void {
    // Clean up all video connections
    this.connections.forEach((_, element) => this.disconnectVideoElement(element));
    this.connections.clear();

    // Stop all streams
    if (this.microphoneStream) {
      this.microphoneStream.getTracks().forEach(track => track.stop());
      this.microphoneStream = null;
    }
    if (this.videoStream) {
      this.videoStream.getTracks().forEach(track => track.stop());
      this.videoStream = null;
    }
  }

  async resumeContext(): Promise<void> {
    if (this.context.state === 'suspended') {
      await this.context.resume();
    }
  }

  getContext(): AudioContext {
    return this.context;
  }
}

export default AudioManager;
