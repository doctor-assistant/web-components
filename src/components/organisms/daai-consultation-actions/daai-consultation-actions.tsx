import { Component, h, Prop, State } from "@stencil/core";
import {
  finishRecording,
  pauseRecording,
  resumeRecording,
  startRecording,
  StartTutorial,
} from "../../../core/Recorder";
import state from "../../../store";
import { getSpecialtiesByProfessionalId } from "../../../utils/indexDb";

@Component({
  tag: "daai-consultation-actions",
  styleUrl: "daai-consultation-actions.css",
  shadow: false,
})
export class DaaiConsultationActions {
  @Prop() apikey: any;
  @Prop() specialty: any;
  @Prop() telemedicine: boolean;
  @Prop() professional: string = "";

  @Prop() success: any;
  @Prop() error: any;
  @Prop() metadata: string;
  @Prop() event: any;

  @State() title: string = "";
  @State() stopAnimation: string = "";

  newRecording() {
    state.status = "initial";
  }

  async choosenMode() {
    if (this.telemedicine) {
      state.status = "choosen";
    } else {
      state.status = "recording";
    }
  }

  choosenSpecialty() {
    state.openModalSpecialty = true;
  }

  async componentDidLoad() {
    const specialtyByProfessionalId = await getSpecialtiesByProfessionalId(
      this.professional
    );
    if (
      specialtyByProfessionalId.mostRecentSpecialty &&
      specialtyByProfessionalId.specialtiesAsStrings
    ) {
      state.specialtyTitle =
        specialtyByProfessionalId.mostRecentSpecialty.title;
      state.chooseSpecialty = specialtyByProfessionalId.mostRecentSpecialty.id;
    }

    const storedValue = localStorage.getItem("checkboxState");
    state.isChecked = storedValue !== null ? JSON.parse(storedValue) : "";
    this.title = `Especialidade`;
  }

  openConfigModal() {
    state.openMenu = true;
    state.isOpenMenuToCancelAnimation = true;
    localStorage.setItem(
      "isOpenMenuToCancelAnimation",
      JSON.stringify(state.isOpenMenuToCancelAnimation)
    );
  }

  renderButtons() {
    switch (state.status) {
      case "initial":
        return (
          <div class="flex items-center justify-center gap-2">
            {!state.defaultSpecialty && (
              <daai-button-with-icon
                title={this.title}
                id="specialty"
                onClick={this.choosenSpecialty}
                disabled={state.defaultSpecialty !== ""}
              >
                {state.specialtyTitle
                  ? state.specialtyTitle
                  : "SOAP Generalista"}
              </daai-button-with-icon>
            )}
            <daai-button-with-icon
              title="Iniciar Registro"
              id="start-recording"
              onClick={() =>
                this.telemedicine ? this.choosenMode() : startRecording(false)
              }
              disabled={!state.microphonePermission}
            >
              <div class="flex items-start justify-start gap-2">
                <daai-mic-icon></daai-mic-icon>
                <daai-text text="Iniciar Registro"></daai-text>
              </div>
            </daai-button-with-icon>
            <daai-button-with-icon
              id={
                localStorage.getItem("isOpenMenuToCancelAnimation")
                  ? "button-menu"
                  : "pulse"
              }
              onClick={this.openConfigModal}
            >
              <daai-menu-icon />
            </daai-button-with-icon>
            {state.openMenu && <daai-config />}
          </div>
        );
      case "choosen":
        return (
          <div class="flex items-center justify-center gap-2">
            <daai-button-with-icon
              title="Iniciar Registro Presencial"
              id="choose-local-consultation"
              onClick={() => startRecording(false)}
            >
              Presencial
            </daai-button-with-icon>
            <daai-button-with-icon
              title="Iniciar Registro Telemedicina"
              id="choose-telemedicine-consultation"
              onClick={() =>
                state.isChecked ? startRecording(true) : StartTutorial()
              }
            >
              Telemedicina
            </daai-button-with-icon>
            <daai-button-with-icon
              title="voltar ao inicio"
              id="button-resume"
              onClick={() => this.newRecording()}
            >
              <daai-resume-recording-icon />
            </daai-button-with-icon>
          </div>
        );
      case "recording":
      case "resume":
        return (
          <div class="flex items-center justify-center gap-2">
            <daai-button-with-icon
              title="Pausar Registro"
              id="pause-recording"
              onClick={() => pauseRecording()}
            >
              <daai-pause-icon />
            </daai-button-with-icon>
            <daai-button-with-icon
              title="Finalizar Registro"
              id="button-finish"
              onClick={() =>
                finishRecording(
                  this.apikey,
                  this.success,
                  this.error,
                  this.specialty,
                  this.metadata,
                  this.event,
                  this.professional
                )
              }
            >
              Finalizar Registro
            </daai-button-with-icon>
          </div>
        );
      case "paused":
        return (
          <div
            class="flex items-center justify-center gap-2
          "
          >
            <daai-button-with-icon
              title="Pausar Registro"
              id="pause-recording-disabled"
              disabled
            >
              <daai-pause-icon />
            </daai-button-with-icon>
            <daai-button-with-icon
              title="Retomar Registro"
              id="start-recording"
              onClick={() => resumeRecording()}
            >
              <div class="flex items-start justify-start gap-2">
                <daai-mic-icon></daai-mic-icon>
                <daai-text text="Retomar"></daai-text>
              </div>
            </daai-button-with-icon>
            <daai-button-with-icon
              title="Retomar Registro"
              id="button-finish"
              onClick={() =>
                finishRecording(
                  this.apikey,
                  this.success,
                  this.error,
                  this.specialty,
                  this.metadata,
                  this.event,
                  this.professional
                )
              }
            >
              Finalizar Registro
            </daai-button-with-icon>
          </div>
        );
      case "upload-ok":
        return (
          <div class="flex items-center justify-center gap-2">
            <daai-button-with-icon
              title="Iniciar Novo Registro"
              id="new-recording"
              onClick={() => this.newRecording()}
            >
              Novo Registro
            </daai-button-with-icon>
          </div>
        );
      default:
        return null;
    }
  }

  render() {
    return <div>{this.renderButtons()}</div>;
  }
}
