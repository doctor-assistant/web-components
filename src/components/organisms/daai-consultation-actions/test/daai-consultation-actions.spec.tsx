import { newSpecPage } from "@stencil/core/testing";
import state from "../../../../store";
import { DaaiConsultationActions } from "../daai-consultation-actions";

describe("daai-consultation-actions", () => {
  it("Verifica se o daai-consultation-actions renderiza corretamente", async () => {
    const page = await newSpecPage({
      components: [DaaiConsultationActions],
      html: `<daai-consultation-actions apikey="test-key" specialty="cardiology" metadata="test-metadata"></daai-consultation-actions>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("chama ChoosenMode e atualiza o estado", async () => {
    const component = new DaaiConsultationActions();
    component.choosenMode();
    expect(state.status).toBe("choosen");
  });

  it("chama ChoosenSpecialty e atualiza o estado", async () => {
    const component = new DaaiConsultationActions();
    component.choosenSpecialty();
    expect(state.openModalSpecialty).toBe(true);
  });

  it("lida com componentDidLoad e atualiza o tÃ­tulo", async () => {
    const component = new DaaiConsultationActions();
    component.specialty = "cardiology";

    const mockGetSpecialtyTitle = jest.fn().mockResolvedValue("Cardiology");
    (global as any).getSpecialtyTitle = mockGetSpecialtyTitle;

    await component.componentDidLoad();
    expect(component.title).toBe("Especialidade Cardiology");
    expect(mockGetSpecialtyTitle).toHaveBeenCalledWith("cardiology");
  });
});
