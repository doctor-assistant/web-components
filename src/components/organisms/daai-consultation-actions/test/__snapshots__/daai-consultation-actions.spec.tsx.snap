// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`daai-consultation-actions Verifica se o daai-consultation-actions renderiza corretamente 1`] = `
<daai-consultation-actions apikey="test-key" metadata="test-metadata" specialty="cardiology">
  <div>
    <div class="flex gap-2 items-center justify-center">
      <daai-button-with-icon id="config-mic" title="Configuração de microfone">
        <daai-config-mic-icon></daai-config-mic-icon>
      </daai-button-with-icon>
      <daai-button-with-icon id="specialty" title="Especialidade null">
        <daai-stethoscope-icon></daai-stethoscope-icon>
      </daai-button-with-icon>
      <daai-button-with-icon disabled="" id="start-recording" title="Iniciar Registro">
        Iniciar Registro
      </daai-button-with-icon>
      <daai-button-with-icon id="button-support" title="Suporte">
        <daai-support-icon></daai-support-icon>
      </daai-button-with-icon>
    </div>
  </div>
</daai-consultation-actions>
`;
